plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts

```diff
@@ -0,0 +1,191 @@
+
+  static create(options: {
+    config: AwsS3Config;
+    integration: AwsS3Integration;
+    logger: Logger;
+    schedule: TaskRunner;
+  }): AwsS3EntityProvider {
+    return new AwsS3EntityProvider(
+      options.config,
+      options.integration,
+      options.logger,
+      options.schedule,
+    );
+  }
+
+  private constructor(
+    private readonly config: AwsS3Config,
+    private readonly integration: AwsS3Integration,
+    logger: Logger,
+    schedule: TaskRunner,
+  ) {
+    this.logger = logger.child({
+      target: this.getProviderName(),
+    });
+
+    this.s3 = new S3({
+      apiVersion: '2006-03-01',
+      credentials: AwsCredentials.create(
+        integration.config,
+        'backstage-aws-s3-provider',
+      ),
+      endpoint: integration.config.endpoint,
+      s3ForcePathStyle: integration.config.s3ForcePathStyle,
+    });
+
+    this.scheduleFn = this.createScheduleFn(schedule);
+  }
+
+  private createScheduleFn(schedule: TaskRunner): () => Promise<void> {
+    return async () => {
+      const taskId = `${this.getProviderName()}:refresh`;
+      return schedule.run({
+        id: taskId,
+        fn: async () => {
+          const logger = this.logger.child({
+            class: AwsS3EntityProvider.prototype.constructor.name,
+            taskId,
+            taskInstanceId: uuid.v4(),
+          });
+
+          try {
+            await this.refresh(logger);
+          } catch (error) {
+            logger.error(error);
+          }
+        },
+      });
+    };
+  }
+
+  /** {@inheritdoc @backstage/plugin-catalog-backend#EntityProvider.getProviderName} */
+  getProviderName(): string {
+    return `awsS3-provider:${this.config.bucket}/${this.config.prefix ?? ''}`;
+  }
+
+  /** {@inheritdoc @backstage/plugin-catalog-backend#EntityProvider.connect} */
+  async connect(connection: EntityProviderConnection): Promise<void> {
+    this.connection = connection;
+    await this.scheduleFn?.();
+  }
+
+  async refresh(logger: Logger) {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    logger.info('Discovering AWS S3 objects');
+
+    const keys = await this.listAllObjectKeys();
+    logger.info(`Discovered ${keys.length} AWS S3 objects`);
+
+    const locations = keys.map(key => this.createLocationSpec(key));
+
+    await this.connection.applyMutation({
+      type: 'full',
+      entities: locations.map(location => {
+        return {
+          locationKey: this.getProviderName(),
+          entity: locationSpecToLocationEntity({ location }),
+        };
+      }),
+    });
+
+    logger.info(`Committed ${locations.length} Locations for AWS S3 objects`);
+  }
+
+  private async listAllObjectKeys(): Promise<string[]> {
+    const keys: string[] = [];
+
+    aws.config.update({ region: this.config.region });
+
+    let continuationToken: string | undefined;
+    let output: ListObjectsV2Output;
+    do {
+      const request = this.s3.listObjectsV2({
```

Comment:
readTree loads all objects, yes.

I thought downloading all objects and keeping the streams and having the buffers for them (even if concurrency is limited) it might be a bit heavy depending on the size of the bucket.
See also `AwsS3DiscoveryProcessor.doRead`.

And then we still need to execute something like the `CatalogProcessorParser` on top to convert the yaml file content into actual entities, etc.

With the approach of using Locations,
- we can spread the load across the cluster
- can refresh individual ones via the UI
- don't need to "reimplement" the logic by AwsS3UrlReader and following processors

The logic is pretty light-weight and could be applied in a similar way for SCM services (e.g., the provider for Bitbucket Cloud, GitHub, etc.).

Reply:
@freben Any thoughts on that still or can I assume this is resolved?

Reply:
@Rugvip Any opinions on this one?

Reply:
I'll resolve this one. I realize that error handling becomes harder to do right too, when just some partial errors in a fraction of the files are found and so on. Unless Rugvip chimes in with something extra, consider this closed

Reply:
Yep thinking that all makes sense @pjungermann. I feel we'd need to power up the entity provider scheduling capabilities if we wanted to have the individual reads happen eagerly @freben.

Something to add into the mix could also be a nicer way for providers to invalidate the entity hash so that they get bumped to the front of the queue if needed, without having to add something like a ETag inside the entity.